### PR TITLE
Shadow plugin

### PR DESCRIPTION
**Pull request by mitchdraft**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/934

### CODE REVIEW

File: projects/gloo/pkg/plugins/shadowing/plugin.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+func getFractionalPercent(numerator uint32) *envoycore.RuntimeFractionalPercent {
+	return &envoycore.RuntimeFractionalPercent{
+		DefaultValue: &envoy_type.FractionalPercent{
+			Numerator:   numerator,
+			Denominator: envoy_type.FractionalPercent_HUNDRED,
+		},
+	}
+}
+
+var (
```
Comment (EItanya) [inline projects/gloo/pkg/plugins/shadowing/plugin.go]: nit: move the errors to the top

File: projects/gloo/api/v1/plugins/shadowing/shadowing.proto [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// If set, Envoy will send a portion of the route's traffic to the shadowed upstream. This can be a useful way to
+// preview a new service's behavior before putting the service in the critical path.
+// Note that this plugin is only applicable to routes with upstream destinations (not redirect or direct response routes).
+// See here for additional information on Envoy's shadowing capabilities: https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto#envoy-api-msg-route-routeaction-requestmirrorpolicy
+message RouteShadowing {
+    // The upstream to which the shadowed traffic should be sent.
+    core.solo.io.ResourceRef upstream = 1;
+
+    // The percent (represented as an integer between 0 and 100) of traffic that should be shadowed.
+    uint32 percent = 2;
```
Comment (yuval-k) [inline projects/gloo/api/v1/plugins/shadowing/shadowing.proto]: this should be float.
similar to Fault plugin

**Comment by mitchdraft** [comment]
Beyond upstreams, destinations include kubernetes services and consul services.
Envoy only allows shadowing to another cluster, identified by it's unique cluster name.
Do kubernetes services and consul services have registered clusters in envoy @yuval-k @ilackarms ?

**Comment by mitchdraft** [comment]
oh it's just `DestinationToUpstreamRef` nice!!

**Comment by mitchdraft** [comment]
in order to use Destination in shadowing.proto we need to move it out of proxy.proto due to a circular dependency:

proxy -> plugins -> shadowing -> proxy

this will be a breaking change and a decent refactor so I'd rather do that in a followup if that's good with you

**Pull request merged**